### PR TITLE
EOS-16880: Extend NTP configuration to storage controllers

### PR DESCRIPTION
**Pull request by 83bhp**
Initial changes:
1. Add new argument (-n|--ntp) to the controller-cli utility
2. Add functionality to configure NTP on the controller

Signed-off-by: Pritam Bhavsar <pritam.bhavsar@seagate.com>

### CODE REVIEW

File: srv/components/controller/files/scripts/controller-cli.sh [hunk 1]
```diff
@@ -155,6 +159,7 @@ help()
     --shutdown-ctrl:- Shutdown the controller (a|b), if second option is ommited
                       then both the controllers will be shutdown
     -l|--load-license:-
+    -m||--ntp        :- configure ntp (user provided server/ip & timezone) on controllers
```
Comment (ypise) [inline srv/components/controller/files/scripts/controller-cli.sh:162]: `-m||--ntp` should be `-n|--ntp`

File: srv/components/controller/files/scripts/controller-cli.sh [hunk 2]
```diff
@@ -443,7 +451,19 @@ parse_args()
...
-
+            -n|--ntp)
+                echo "parse_args(): configure ntp" >> $logfile
+                if [ -z "$2" ]; then
+                    echo "Error: NTP server not provided" && exit 1;
+                fi
+                if [ -z "$3" ]; then
+                    echo "Error: Timezone not provided" && exit 1;
+                fi
+                ntp_server="$2"
```
Comment (ypise) [inline srv/components/controller/files/scripts/controller-cli.sh:462]: This parameter should come from pillar data:
`system/ntp/time_server`

File: srv/components/controller/files/scripts/controller-cli.sh [hunk 3]
```diff
@@ -443,7 +451,19 @@ parse_args()
...
+                ntp_tz="$3"
```
Comment (ypise) [inline srv/components/controller/files/scripts/controller-cli.sh:463]: This parameter should come from pillar data:
`system/ntp/time_zone`

File: srv/components/controller/files/scripts/provision.sh [hunk 4]
```diff
@@ -1747,3 +1747,73 @@ do_provision()
             echo "No provisioning details found on the controller"
     }
 }
+
+ntp_status_get()
+{
+    ## objects name in the xml
+    _xml_obj_bt="ntp-status"
+    # Property value to be extracted from xml
+    _xml_obj_plist=("ntp-status" "ntp-server-address" "ntp-contact-time")
```
Comment (ypise) [inline srv/components/controller/files/scripts/provision.sh:1756]: Why not read JSON output?

File: srv/components/controller/files/scripts/provision.sh [hunk 5]
```diff
@@ -1747,3 +1747,73 @@ do_provision()
...
+    _tmp_file="$tmpdir/ntp"
+    _ntpinfo="$tmpdir/ntp_info"
+    [ -f $_tmp_file ] && rm -rf $_tmp_file
+    [ -f $_ntpinfo ] && rm -rf $_ntpinfo
+
+    _cmd="show ntp-status"
+    echo "ntp_status_get(): Entry" >> $logfile
+    echo "ntp_status_get(): running command: '$_cmd'" >> $logfile
+    cmd_run "$_cmd" 
+    parse_xml $xml_doc $_xml_obj_bt "${_xml_obj_plist[@]}" > $_tmp_file
```
Comment (ypise) [inline srv/components/controller/files/scripts/provision.sh:1767]: Would be better to parse JSON output.

**Review by ypise — CHANGES_REQUESTED** [review]

File: srv/components/controller/files/scripts/controller-cli.sh (hunk 1 above)
Comment (83bhp) [reply srv/components/controller/files/scripts/controller-cli.sh:162]: yes, taken care of.

File: srv/components/controller/files/scripts/controller-cli.sh (hunk 2 above)
Comment (83bhp) [reply srv/components/controller/files/scripts/controller-cli.sh:462]: yes, this script is supposed to be called from api which is supposed to read time_server from pillar & provide it to this utility.

File: srv/components/controller/files/scripts/provision.sh (hunk 4 above)
Comment (83bhp) [reply srv/components/controller/files/scripts/provision.sh:1756]: the utility already has implementation for the xml parseer

File: srv/components/controller/files/scripts/provision.sh (hunk 5 above)
Comment (83bhp) [reply srv/components/controller/files/scripts/provision.sh:1767]: I have not written any parse, the utility already has implementation for the xml parser.

File: srv/components/controller/files/scripts/controller-cli.sh (hunk 3 above)
Comment (83bhp) [reply srv/components/controller/files/scripts/controller-cli.sh:463]: yes, this script is supposed to be called from api which is supposed to read time_server from pillar & provide it to this utility

**Review by ypise — APPROVED** [review]

**Review by ypise — APPROVED** [review]

**Pull request merged**